src/python/pants/engine/target.py

```diff
@@ -677,6 +672,7 @@ def __eq__(self, other: Any) -> bool:
         return (
             self._hashcode == other._hashcode
             and self.members == other.members
+            # TODO: Use a recursive memoized __eq__ if this ever shows up in profiles.
```

Comment:
It might also be good for these two classes to mention one another: if we do end up extracting this, we'll want to reuse it across them I think.